File: benchmarks/typescript/bench-persistence.ts
// Persistence + memory benchmark — TypeScript binding.

import { performance } from "node:perf_hooks";
import { mkdtempSync, statSync, unlinkSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import Database from "better-sqlite3";
import {
  Smritidb,
  openPersistentStore,
  persistStore,
  sqliteAdapter,
} from "@tanvrit/smritidb";

import { D, PERSIST_N, WAL_APPENDS, WARMUP_ITERS } from "./bench-utils.js";

export interface PersistOp {
  n: number;
  bytes: number;
  wall_seconds: number;
  mb_per_sec: number;
}

export interface WalResult {
  n: number;
  wall_seconds: number;
  ops_per_sec: number;
  note?: string;
}

export interface PersistenceOut {
  snapshot_write: PersistOp;
  snapshot_read: PersistOp;
  wal_append: WalResult;
}

function populate(store: Smritidb, n: number): void {
  for (let i = 0; i < n; i++) {
    store.put(`item_${i}`, `value for item ${i}`);
  }
}

function fileSize(path: string): number {
  try {
    return statSync(path).size;
  } catch {
    return 0;
  }
}

function cleanupFile(path: string) {
  for (const suffix of ["", "-journal", "-wal", "-shm"]) {
    try {
      unlinkSync(path + suffix);
    } catch {
      // ignore
    }
  }
}

export async function runPersistence(): Promise<PersistenceOut> {
  console.log("[3/4] persistence (TypeScript)");

  const dir = mkdtempSync(join(tmpdir(), "smritidb-bench-"));
  const writePath = join(dir, "write.db");
  const readPath = join(dir, "read.db");

  // -- snapshot_write --
  cleanupFile(writePath);
  const writeStore = new Smritidb({ dimension: D });
  populate(writeStore, PERSIST_N);
  const writeDb = new Database(writePath);
  const writeAdapter = sqliteAdapter(writeDb);
  const tw0 = performance.now();
  await persistStore(writeStore, writeAdapter);
  const writeWall = (performance.now() - tw0) / 1000;
  await writeAdapter.close();
  const writeBytes = fileSize(writePath);
  const writeMbs = writeBytes / 1_000_000 / writeWall;
  console.log(
    `  snapshot_write N=${PERSIST_N} bytes=${writeBytes} wall=${writeWall.toFixed(
      3,
    )}s = ${writeMbs.toFixed(1)} MB/s`,
  );

  // -- snapshot_read --
  cleanupFile(readPath);
  const seedStore = new Smritidb({ dimension: D });
  populate(seedStore, PERSIST_N);
  const seedDb = new Database(readPath);
  const seedAdapter = sqliteAdapter(seedDb);
  await persistStore(seedStore, seedAdapter);
  await seedAdapter.close();
  const readBytes = fileSize(readPath);

  for (let i = 0; i < WARMUP_ITERS; i++) {
    const db = new Database(readPath, { readonly: true });
    const ad = sqliteAdapter(db);
    await openPersistentStore({ dimension: D, adapter: ad });
    await ad.close();
  }

  const tr0 = performance.now();
  const readDb = new Database(readPath, { readonly: true });
  const readAdapter = sqliteAdapter(readDb);
  const restored = await openPersistentStore({ dimension: D, adapter: readAdapter });
  const readWall = (performance.now() - tr0) / 1000;
  if (restored.size() !== PERSIST_N) {
    throw new Error(`restored.size() = ${restored.size()}, expected ${PERSIST_N}`);
  }
  await readAdapter.close();
  const readMbs = readBytes / 1_000_000 / readWall;
  console.log(
    `  snapshot_read  N=${PERSIST_N} bytes=${readBytes} wall=${readWall.toFixed(
      3,
    )}s = ${readMbs.toFixed(1)} MB/s`,
  );

  // -- wal_append (surrogate: TS adapter exposes write/read but no
  // append_wal on the public surface; persist() rewrites the snapshot.
  // Footnote in RESULTS.md.).
  const walPath = join(dir, "wal.db");
  cleanupFile(walPath);
  const walDb = new Database(walPath);
  const walAdapter = sqliteAdapter(walDb);
  const walStore = new Smritidb({ dimension: D });
  walStore.put("seed", "x");

  for (let i = 0; i < WARMUP_ITERS; i++) {
    await persistStore(walStore, walAdapter);
  }
  const tWal0 = performance.now();
  for (let i = 0; i < WAL_APPENDS; i++) {
    await persistStore(walStore, walAdapter);
  }
  const walWall = (performance.now() - tWal0) / 1000;
  const walOps = WAL_APPENDS / walWall;
  await walAdapter.close();
  console.log(
    `  wal_append*   N=${WAL_APPENDS} wall=${walWall.toFixed(3)}s = ${walOps.toFixed(
      0,
    )} ops/s (surrogate: repeated persistStore())`,
  );

  return {
    snapshot_write: {
      n: PERSIST_N,
      bytes: writeBytes,
      wall_seconds: writeWall,
      mb_per_sec: writeMbs,
    },
    snapshot_read: {
      n: PERSIST_N,
      bytes: readBytes,
      wall_seconds: readWall,
      mb_per_sec: readMbs,
    },
    wal_append: {
      n: WAL_APPENDS,
      wall_seconds: walWall,
      ops_per_sec: walOps,
      note: "TS adapter exposes write/read; this measures repeated full persistStore() instead.",
    },
  };
}

export function runMemory(): {
  n_items: number;
  rss_before_bytes: number;
  rss_after_bytes: number;
  bytes_per_item: number;
  theoretical_floor_bytes_per_item: number;
} {
  console.log("[4/4] memory (TypeScript)");
  // V8 schedules background GC asynchronously, so process.rss is noisy.
  // We prefer heapUsed when --expose-gc is available (run-all.sh passes it);
  // otherwise we fall back to rss, which can swing wildly. To make the
  // benchmark useful even without --expose-gc we synthesize the floor from
  // the dominant cost: PERSIST_N × D bytes of Uint8Array key data.
  const hasGc = typeof (globalThis as { gc?: () => void }).gc === "function";
  if (hasGc) (globalThis as { gc: () => void }).gc();
  const before = hasGc ? process.memoryUsage().heapUsed : process.memoryUsage().rss;
  const store = new Smritidb({ dimension: D });
  for (let i = 0; i < PERSIST_N; i++) {
    store.put(`item_${i}`, `value for item ${i}`);
  }
  if (hasGc) (globalThis as { gc: () => void }).gc();
  const after = hasGc ? process.memoryUsage().heapUsed : process.memoryUsage().rss;
  const delta = Math.max(0, after - before);
  const perItem = delta / PERSIST_N;
  console.log(
    `  ${hasGc ? "heapUsed" : "rss"} before=${before}B after=${after}B ` +
      `delta=${delta}B = ${perItem.toFixed(1)} bytes/item ` +
      `${hasGc ? "" : "(noisy without --expose-gc)"}`,
  );
  // Keep store alive past the sample.
  void store.size();
  return {
    n_items: PERSIST_N,
    rss_before_bytes: before,
    rss_after_bytes: after,
    bytes_per_item: perItem,
    theoretical_floor_bytes_per_item: D / 8,
  };
}

if (import.meta.url === `file://${process.argv[1]}`) {
  runPersistence().then((p) => {
    console.log(JSON.stringify({ persistence: p, memory: runMemory() }, null, 2));
  });
}

File: benchmarks/python/bench_persistence.py
#!/usr/bin/env python3
"""Persistence throughput benchmark — Python binding.

Measures SQLite snapshot write/read MB/s and (where the Python binding
exposes it) WAL append ops/s. The Python `PersistentStore` only exposes
`persist()`, which writes a full snapshot — there's no direct WAL append
API on the Python surface today. We fall back to repeated `persist()`
calls and surface that as a snapshot-rewrite-rate so the table has a
number to compare against the other bindings.
"""

from __future__ import annotations

import os
import resource
import tempfile
import time
from typing import Any

from smritidb import PersistentStore

D = 10_000
PERSIST_N = 10_000
WAL_APPENDS = 1_000
WARMUP_ITERS = 3


def _populate(store: PersistentStore, n: int) -> None:
    for i in range(n):
        store.put(f"item_{i}", f"value for item {i}".encode())


def _cleanup(path: str) -> None:
    for suffix in ("", "-journal", "-wal", "-shm"):
        p = path + suffix
        if os.path.exists(p):
            os.unlink(p)


def run() -> dict[str, Any]:
    print("[3/4] persistence (Python)")

    with tempfile.TemporaryDirectory() as tmp:
        write_path = os.path.join(tmp, "write.db")
        read_path = os.path.join(tmp, "read.db")
        wal_path = os.path.join(tmp, "wal.db")

        # -- snapshot_write --
        _cleanup(write_path)
        store = PersistentStore.open_sqlite(write_path, dimension=D)
        _populate(store, PERSIST_N)
        t0 = time.perf_counter()
        store.persist()
        write_wall = time.perf_counter() - t0
        store.close()
        write_bytes = os.path.getsize(write_path)
        write_mbs = (write_bytes / 1_000_000) / write_wall
        print(
            f"  snapshot_write N={PERSIST_N} bytes={write_bytes} "
            f"wall={write_wall:.3f}s = {write_mbs:.1f} MB/s"
        )

        # -- snapshot_read --
        _cleanup(read_path)
        store_r = PersistentStore.open_sqlite(read_path, dimension=D)
        _populate(store_r, PERSIST_N)
        store_r.persist()
        store_r.close()
        read_bytes = os.path.getsize(read_path)

        for _ in range(WARMUP_ITERS):
            s = PersistentStore.open_sqlite(read_path)
            s.close()

        t0 = time.perf_counter()
        s = PersistentStore.open_sqlite(read_path)
        read_wall = time.perf_counter() - t0
        assert s.size() == PERSIST_N
        s.close()
        read_mbs = (read_bytes / 1_000_000) / read_wall
        print(
            f"  snapshot_read  N={PERSIST_N} bytes={read_bytes} "
            f"wall={read_wall:.3f}s = {read_mbs:.1f} MB/s"
        )

        # -- wal_append (degraded — Python binding does not expose append_wal
        # directly; surrogate uses repeated persist() at small N for
        # rate-of-snapshots-per-sec). Document this in RESULTS.md footnote.
        _cleanup(wal_path)
        wal_store = PersistentStore.open_sqlite(wal_path, dimension=D)
        wal_store.put("seed", b"x")  # small steady-state store
        for _ in range(WARMUP_ITERS):
            wal_store.persist()
        t0 = time.perf_counter()
        for _ in range(WAL_APPENDS):
            wal_store.persist()
        wal_wall = time.perf_counter() - t0
        wal_ops = WAL_APPENDS / wal_wall
        wal_store.close()
        print(
            f"  wal_append*   N={WAL_APPENDS} wall={wal_wall:.3f}s = {wal_ops:.0f} ops/s "
            "(surrogate: repeated persist())"
        )

    return {
        "snapshot_write": {
            "n": PERSIST_N,
            "bytes": write_bytes,
            "wall_seconds": write_wall,
            "mb_per_sec": write_mbs,
        },
        "snapshot_read": {
            "n": PERSIST_N,
            "bytes": read_bytes,
            "wall_seconds": read_wall,
            "mb_per_sec": read_mbs,
        },
        "wal_append": {
            "n": WAL_APPENDS,
            "wall_seconds": wal_wall,
            "ops_per_sec": wal_ops,
            "note": "Python binding has no direct append_wal; this measures repeated full persist() instead.",
        },
    }


def _rss_bytes() -> int:
    """Best-effort current process RSS in bytes.

    `ru_maxrss` from resource is a *high-water mark* (never shrinks), which
    makes delta measurements unreliable. We shell out to `ps` for current
    RSS on macOS, parse /proc/self/status on Linux, and fall back to
    `ru_maxrss` elsewhere.
    """
    import os
    import subprocess
    import sys

    if sys.platform == "darwin":
        try:
            out = subprocess.check_output(
                ["ps", "-o", "rss=", "-p", str(os.getpid())], text=True
            ).strip()
            return int(out) * 1024
        except Exception:
            pass
    elif sys.platform.startswith("linux"):
        try:
            with open("/proc/self/status") as f:
                for line in f:
                    if line.startswith("VmRSS:"):
                        return int(line.split()[1]) * 1024
        except Exception:
            pass
    # Fallback: ru_maxrss. Scale: bytes on macOS, KiB on Linux.
    scale = 1 if sys.platform == "darwin" else 1024
    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss * scale


def run_memory() -> dict[str, Any]:
    """Insert PERSIST_N items into an in-memory Store, measure RSS delta.

    Caveat on macOS: the system allocator (malloc + memory compressor) tends
    to keep RSS flat across small allocations — the 10000×1250 B ≈ 12 MB
    store often fits inside the interpreter's pre-grown arena. We still
    report the delta but a 0 reading on macOS doesn't mean "zero overhead";
    it means "below RSS measurement resolution". The Rust runner has the
    same limitation. The honest per-item floor is calculated by counting
    the dominant data structure: a `bytes` of D bytes per item.
    """
    print("[4/4] memory (Python)")
    import gc
    gc.collect()
    before = _rss_bytes()
    from smritidb import Store
    store = Store(dimension=D)
    for i in range(PERSIST_N):
        store.put(f"item_{i}", f"value for item {i}")
    gc.collect()
    after = _rss_bytes()
    delta = max(0, after - before)
    per_item = delta / PERSIST_N
    # Synthetic floor: every item carries a D-byte key bytes object,
    # a value, a UUID, plus dict / HashMap overhead. We surface this
    # alongside the noisy RSS measurement so the table has *something*
    # meaningful when the OS allocator is compressing under us.
    synthetic = float(D)  # at minimum, the key
    print(
        f"  RSS before={before}B after={after}B delta={delta}B = {per_item:.1f} bytes/item "
        f"(synthetic floor: {synthetic:.0f} B/item)"
    )
    # Keep store alive past the sample.
    _ = store.size()
    return {
        "n_items": PERSIST_N,
        "rss_before_bytes": before,
        "rss_after_bytes": after,
        "bytes_per_item": per_item,
        "synthetic_floor_bytes_per_item": synthetic,
        "theoretical_floor_bytes_per_item": D // 8,
    }


if __name__ == "__main__":
    import json, sys
    out = {"persistence": run(), "memory": run_memory()}
    json.dump(out, sys.stdout, indent=2)
    print()
